common/cpp/src/google_smart_card_common/value_unittest.cc

```diff
@@ -0,0 +1,339 @@
+  const Value value(123);
+  EXPECT_EQ(value.type(), Value::Type::kInteger);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_TRUE(value.is_integer());
+  EXPECT_FALSE(value.is_float());
+  EXPECT_FALSE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_EQ(value.GetInteger(), 123);
+  EXPECT_DOUBLE_EQ(value.GetFloat(), 123);
+}
+
+TEST(ValueTest, IntegerDefault) {
+  const Value value(Value::Type::kInteger);
+  EXPECT_EQ(value.type(), Value::Type::kInteger);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_TRUE(value.is_integer());
+  EXPECT_FALSE(value.is_float());
+  EXPECT_FALSE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_EQ(value.GetInteger(), 0);
+  EXPECT_DOUBLE_EQ(value.GetFloat(), 0);
+}
+
+TEST(ValueTest, Float) {
+  const Value value(123.456);
+  EXPECT_EQ(value.type(), Value::Type::kFloat);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_FALSE(value.is_integer());
+  EXPECT_TRUE(value.is_float());
+  EXPECT_FALSE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_DOUBLE_EQ(value.GetFloat(), 123.456);
+}
+
+TEST(ValueTest, FloatDefault) {
+  const Value value(Value::Type::kFloat);
+  EXPECT_EQ(value.type(), Value::Type::kFloat);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_FALSE(value.is_integer());
+  EXPECT_TRUE(value.is_float());
+  EXPECT_FALSE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_DOUBLE_EQ(value.GetFloat(), 0);
+}
+
+TEST(ValueTest, String) {
+  const Value value("foo");
+  EXPECT_EQ(value.type(), Value::Type::kString);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_FALSE(value.is_integer());
+  EXPECT_FALSE(value.is_float());
+  EXPECT_TRUE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_EQ(value.GetString(), "foo");
+}
+
+TEST(ValueTest, StringDefault) {
+  const Value value(Value::Type::kString);
+  EXPECT_EQ(value.type(), Value::Type::kString);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_FALSE(value.is_integer());
+  EXPECT_FALSE(value.is_float());
+  EXPECT_TRUE(value.is_string());
+  EXPECT_FALSE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_EQ(value.GetString(), "");
+}
+
+TEST(ValueTest, Binary) {
+  const std::vector<uint8_t> bytes = {1, 2, 3};
+  const Value value(bytes);
+  EXPECT_EQ(value.type(), Value::Type::kBinary);
+  EXPECT_FALSE(value.is_null());
+  EXPECT_FALSE(value.is_integer());
+  EXPECT_FALSE(value.is_float());
+  EXPECT_FALSE(value.is_string());
+  EXPECT_TRUE(value.is_binary());
+  EXPECT_FALSE(value.is_dictionary());
+  EXPECT_FALSE(value.is_array());
+  EXPECT_EQ(value.GetBinary(), bytes);
```

Comment:
Does this work as expected (i.e., does it actually compare the contents of the arrays)?

Reply:
Yes. Verified this manually by comparing with wrong bytes - it failed with:
```
../../src/google_smart_card_common/value_unittest.cc:177: Value of: bytes2
  Actual: { '\x1' (1), '\x2' (2), '\x3' (3), '\x4' (4) }
Expected: value.GetBinary()
Which is: { '\x1' (1), '\x2' (2), '\x3' (3) }
```